### DISCUSSION TITLE
Java config evaluation at compile time

### DISCUSSION THREAD

**Discussion by dev92** [apple/pkl#277] [Q&A]
Hi,
From the java code-gen example I see we generate java classes from pkl template and then during runtime we evaluate with a pkl file that has the actual values.
Is it possible to check pkl file during compilation phase in Java?, instead of getting runtime errors

**Comment by edward3h** [comment]
I did this by adding an evaluator in my Gradle configuration. 
Example:

```
pkl {
    javaCodeGenerators {
        register("genJava") {
            generateGetters = true
            sourceModules.add(file("src/main/resources/ConfigurationTemplate.pkl"))
        }
    }
    evaluators {
        register("evalDev") {
            sourceModules.add(file("src/main/resources/dev.pkl"))
            transitiveModules.from(file("src/main/resources/ConfigurationTemplate.pkl"))
            outputFile = layout.buildDirectory.file("evaldev.pcf")
        }
    }
}
```

When my `dev.pkl` file does not set a value for a property defined in `ConfigurationTemplate.pkl`, this fails the Gradle build.

**Reply by holzensp** [reply]
This is the solution I would have proposed also; ~although... if your scenario is entirely build-time, you could even remove your (run-time) Pkl dependency and just read the rendered JSON.~ (edit: since you’re using codegen, there are risks going via a less expressive format, because of e.g. inheritance and/or non-JSON types)

**Reply by bioball** [reply]
I would recommend setting up a Java unit test to evaluate Pkl. This would be the most representative of how you will be using Pkl in production.

Adding evaluation via Gradle is pretty good, but those are textual evaluators, and they are slightly different than how Java evaluates Pkl.

Our textual evaluators evaluate the `output.text` (or `output.files`) property of a Pkl program, while our Java evaluator's [`evaluate` method](https://javadocs.dev/org.pkl-lang/pkl-config-java/0.25.2/org/pkl/config/java/ConfigEvaluator.html#evaluate(org.pkl.core.ModuleSource)) evaluates `module` (the top-level object).

For example, this Pkl code would pass if run through our `evaluators` extension, but would fail through our Java-side `evaluate()`.
The `evaluators` extension will pass because it will evaluate to the string "I am totally okay". OTOH, the Java-side `evaluate()` will fail because it evaluates `module`. And `module` is not valid because property `foo` is not set.

```groovy
foo: String

output {
  text = "I am totally okay"
}
```
